internal/util/dependency/deletion_guard.go

```diff
@@ -82,13 +82,34 @@ func AddDeletionGuard[objTP objectType[objT], objT any, depTP objectType[depT],
 		if err != nil {
 			return reconcile.Result{}, nil
 		}
-		if len(refObjects) == 0 {
-			log.V(4).Info("Removing finalizer")
-			patch := finalizers.RemoveFinalizerPatch(dep)
-			return ctrl.Result{}, k8sClient.Patch(ctx, dep, patch, client.ForceOwnership, fieldOwner)
+
+		depKind := dep.GetObjectKind().GroupVersionKind().Kind
+		depUID := dep.GetUID()
+		isOwned := func(obj objTP) bool {
+			owners := obj.GetOwnerReferences()
+			for i := range owners {
+				owner := &owners[i]
+				if owner.Kind == depKind && owner.UID == depUID {
+					return true
+				}
+			}
+			return false
 		}
-		log.V(5).Info("Waiting for dependencies", "dependencies", len(refObjects))
-		return ctrl.Result{}, nil
+
+		// Don't proceed if there are any referring objects, except owners of this object.
+		// We don't block the deletion of the object which created us, because
+		// that would cause a deadlock.
+		for i := range refObjects {
+			refObject := &refObjects[i]
+			if !isOwned(refObject) {
```

Comment:
I'm having difficulties understanding the direction of the ownership relationship. I'm writing this down for my own understanding.

In the subnet/network example, `dep` represents a network object, while `refObjects` represents all objects that have a dependency on it, possibly including a subnet object.

`isOwned(obj)` checks if `dep` owns `obj`, `obj` being one of the `refObjects`. It other words, it checks if `dep` created `obj`.

We don't remove the finalizer until `refObjects` is empty (no other object depend on this resource), or if every object in `refObjects` were created by the resource itself.

I think it would be easier to understand if we rename `isOwned()` to `owns()`.

Reply:
`owns` and `isOwned` have opposite meanings, though. In this case `isOwned` is correct: we are checking the owner references of the object to see if it has the specific owner, i.e. it is owned by that object.

I take your point on the naming of things, btw. I haven't yet come up with anything unambiguous, though, so for now I think the best we can do is to be consistent. This is why I reversed the naming in the previous PR: previously it wasn't consistent and that was impossibly hard to understand.

Reply:
We check the owner reference of `obj` to see if `dep` owns `obj`, so calling `owns(obj)` is a lot less confusing IMO.